app/ui/browser-modern/sagas/session-sagas.js

```diff
@@ -0,0 +1,91 @@
+specific language governing permissions and limitations under the License.
+*/
+
+import { takeLatest, throttle } from 'redux-saga';
+import { race, call, take, put, select } from 'redux-saga/effects';
+
+import { ipcRenderer } from '../../../shared/electron';
+import { wrapped } from './helpers';
+import { getWindowId } from '../selectors/ui';
+import { serializeAppState } from '../util/session-util';
+import * as ActionTypes from '../constants/action-types';
+import * as EffectTypes from '../constants/effect-types';
+import * as SessionEffects from '../actions/session-effects';
+
+const THROTTLE_DURATION = 1000; // ms
+
+export default function*() {
+  yield [
+    takeLatest(...wrapped(EffectTypes.GET_SESSION_KEY, getSessionKey)),
+    takeLatest(...wrapped(EffectTypes.SET_SESSION_KEY, setSessionKey)),
+    takeLatest(...wrapped(EffectTypes.DEL_SESSION_KEY, delSessionKey)),
+  ];
+
+  // Make sure we're cancelling the `watchAppStateChanges` task when either
+  // closing or reloading browser windows, so that everything is aborted even
+  // if recent actions cause the `saveAppStateToSession` task to start running
+  // due to throttling.
+  const winner = yield race({
+    task: call(watchAppStateChanges, saveAppStateToSession),
+    reload: take(EffectTypes.RELOAD_WINDOW),
+    close: take(EffectTypes.CLOSE_WINDOW),
+  });
+
+  if (winner.close) {
+    // When closing a browser window, remove it from the session.
+    yield deleteAppStateFromSession();
+  } else if (winner.reload) {
+    // When reloading a browser window, save it one last time to the session,
+    // so that everything is guaranteed to be up to date. Otherwise, if the
+    // state changes just before throttling finishes, the very latest changes
+    // won't persist.
+    yield saveAppStateToSession();
+  }
+}
+
+function* getSessionKey({ key }) {
+  ipcRenderer.send('session:getkey', key);
+}
+
+function* setSessionKey({ key, value }) {
+  ipcRenderer.send('session:setkey', key, value);
+}
+
+function* delSessionKey({ key }) {
+  ipcRenderer.send('session:delkey', key);
+}
+
+function* watchAppStateChanges(fn) {
+  // It's safe to watch all actions here since we're throttling and not just
+  // debouncing (potentially indefinitely if actions are dispatched constantly
+  // with a lower frequency than the debounce delay). No reason to watch effects,
+  // which are defined in `../constants/effect-types`, since they always
+  // dispatch actions in `../constants/action-types` when having to touch the
+  // app state.
+  const pattern = Object.values(ActionTypes);
+  yield* throttle(THROTTLE_DURATION, ...wrapped(pattern, fn));
+}
+
+function* saveAppStateToSession() {
+  // Invoked periodically when changes are made to the app state.
+  const store = yield select(serializeAppState);
+  const windowId = yield select(getWindowId);
+  yield put(SessionEffects.setSessionKey(['browserWindows', windowId], store));
```

Comment:
I'll add comments where the API is defined. Currently the `windowId` is top level, part of the app state itself.